### PR TITLE
Fix quantizer propagation algo

### PR DESCRIPTION
**Pull request by asenina**
### Related tickets
#60615

### Tests
Tests were added in `test_quantizer_propagation_solver.py`

### CODE REVIEW

File: nncf/common/quantization/quantizer_propagation/graph.py [hunk 1]
```diff
@@ -53,6 +53,7 @@
 from nncf.common.quantization.structs import UnifiedScaleType
 from nncf.common.utils.helpers import should_consider_scope
 from nncf.common.utils.logger import logger as nncf_logger
+from nncf.common.graph.operator_metatypes import OUTPUT_NOOP_METATYPES
```
Comment (alexsu52) [inline nncf/common/quantization/quantizer_propagation/graph.py:56]: Please, optimize imports https://www.jetbrains.com/pycharm/guide/tips/optimize-imports/

**Review by alexsu52 — APPROVED** [review]
Looks good to me.

File: tests/torch/quantization/test_quantizer_propagation_solver.py [hunk 2]
```diff
@@ -897,13 +901,17 @@ def test_merged_qconfig_list_is_independent_of_branch_qconfig_list_order(self,
                                                 'target_branching_node_for_primary_quantizer',
                                                 'expected_status'))
 
+    InitNodeTestStruct = namedtuple('InitNodeTestStruct',
```
Comment (vshampor) [inline tests/torch/quantization/test_quantizer_propagation_solver.py:904]: Please don't use namedtuples.

**Review by vshampor — CHANGES_REQUESTED** [review]
The PR description is missing. Also missing are the tests for the "quantize_outputs": true vs false situations, and the logic handling default quantization for outputs that weren't quantized as a result of the quantizer propagation, but should be for "quantize_outputs": true.

File: tests/torch/quantization/test_quantizer_propagation_solver.py (hunk 2 above)
Comment (asenina) [reply tests/torch/quantization/test_quantizer_propagation_solver.py:904]: Just for my understanding, what significant advantages will I get from replacing a `namedtuple` with a `class`?

Comment (vshampor) [reply tests/torch/quantization/test_quantizer_propagation_solver.py:904]: You get better IDE completion, refactoring, type hints and the ability to do `isinstance` checks on corresponding objects with the specific class name instead of the generic `NamedTuple`. You lose in performance, which is not that relevant here; and even performance is better achieved using `__slots__` in regular classes rather than namedtuples.

**Comment by asenina** [comment]
Jenkins please retry a build

**Comment by ljaljushkin** [comment]
Jenkins please retry a build

**Comment by asenina** [comment]
Jenkins please retry a build

File: nncf/common/quantization/quantizer_propagation/graph.py (hunk 1 above)
Comment (asenina) [reply nncf/common/quantization/quantizer_propagation/graph.py]: Done

**Review by vshampor — CHANGES_REQUESTED** [review]
I would also like to see some visualizations of the impacted reference .dot graphs before and after the change - pretty hard to parse the .dot changes visually.

File: nncf/common/insertion_point_graph.py [hunk 3]
```diff
@@ -120,9 +120,12 @@ def __init__(self, nncf_graph: NNCFGraph, weight_modifiable_node_names: List[NNC
         for edge in self._base_nx_graph.edges:
             input_port_id = self._base_nx_graph.edges[edge][NNCFGraph.INPUT_PORT_ID_EDGE_ATTR]
             dtype = self._base_nx_graph.edges[edge][NNCFGraph.DTYPE_EDGE_ATTR]
+            multiplicity_edge = self._base_nx_graph.edges[edge][NNCFGraph.MULTIPLICITY_EDGE_ATTR]
             from_node, to_node = edge
+
             attrs = {INPUT_PORT_ID: input_port_id,
-                     self.IS_INTEGER_PATH_EDGE_ATTR: dtype is Dtype.INTEGER}
+                     self.IS_INTEGER_PATH_EDGE_ATTR: dtype is Dtype.INTEGER,
+                     NNCFGraph.MULTIPLICITY_EDGE_ATTR: multiplicity_edge}
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/common/insertion_point_graph.py:128]: This is not a temporary attribute. Its addition makes it easier to work with the IP graph. I believe that for the correct processing of any of the graphs, we need to have this attribute.

File: nncf/common/insertion_point_graph.py [hunk 4]
```diff
@@ -141,7 +144,12 @@ def __init__(self, nncf_graph: NNCFGraph, weight_modifiable_node_names: List[NNC
                 pre_hook_ips = list(target_node_name_vs_pre_hook_ips[original_node.node_name])
                 pre_hook_ips = sorted(pre_hook_ips, key=lambda x: x.input_port_id)
                 in_edges = list(self.in_edges(operator_node_key))
-                input_port_id_vs_edge = {self.edges[edge][INPUT_PORT_ID]: edge for edge in in_edges}
+                input_port_id_vs_edge = {}
+                for edge in in_edges:
+                    input_port_id = self.edges[edge][INPUT_PORT_ID]
+                    for i in range(self.edges[edge][NNCFGraph.MULTIPLICITY_EDGE_ATTR]):
+                        input_port_id_vs_edge[input_port_id - i] = edge
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/common/insertion_point_graph.py:151]: It really doesn't matter.

File: nncf/common/graph/graph.py [hunk 5]
```diff
@@ -110,7 +110,8 @@ def __init__(self, from_node: NNCFNode, to_node: NNCFNode,
                  input_port_id: int,
                  output_port_id: int,
                  tensor_shape: List[int],
-                 dtype: Dtype):
+                 dtype: Dtype,
+                 multiplicity_edge: int):
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/common/graph/graph.py]: Done

File: nncf/common/quantization/quantizer_propagation/graph.py [hunk 6]
```diff
@@ -751,6 +752,34 @@ def traverse_fn(curr_node_key: str, all_pqs: Set[PropagatingQuantizer]) -> \
         self.traverse_graph(node_key, traverse_fn, retval)
         return retval
 
+    def check_status_for_output_nodes(self, from_node_key: str) -> 'TransitionStatus':
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/common/quantization/quantizer_propagation/graph.py]: Done

File: nncf/common/quantization/quantizer_propagation/graph.py [hunk 7]
```diff
@@ -751,6 +752,34 @@ def traverse_fn(curr_node_key: str, all_pqs: Set[PropagatingQuantizer]) -> \
...
+        """
+        Checks that all branches outgoing from the branching node can be quantized
+        (They do not contain an output that should not be quantized).
+        """
+        from nncf.common.quantization.quantizer_propagation.solver import TransitionStatus
+
+        def traverse_fn(curr_node_key: str) -> TransitionStatus:
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/common/quantization/quantizer_propagation/graph.py]: Done

**Comment by wonjuleee** [comment]
test this please

File: nncf/common/insertion_point_graph.py [hunk 8]
```diff
@@ -154,8 +162,9 @@ def __init__(self, nncf_graph: NNCFGraph, weight_modifiable_node_names: List[NNC
                     }
 
                     self.add_node(ip_node_key, **pre_hook_ip_attrs)
-
-                    self.remove_edge(from_node_key, to_node_key)
+                    self.edges[edge][NNCFGraph.EDGE_MULTIPLICITY_ATTR] -=1
```
Comment (vinnamkim) [inline nncf/common/insertion_point_graph.py:165]:
```suggestion
                    self.edges[edge][NNCFGraph.EDGE_MULTIPLICITY_ATTR] -= 1
```

**Comment by vshampor** [comment]
The newly added test_intermediate_output before and after the fix:

*Before:*
before

*After:*
after


Note that the intermediate output is no longer quantized since the default behaviour is to have `"quantize_outputs": False`, while still having activations of both convs quantized.

**Review by vshampor — APPROVED** [review]

**Comment by vshampor** [comment]
Hangs on test_quantize_network[w_sym_t_a_sym_t-mask_rcnn], probably due to excessive complexity of the algo implementation. Will try to optimize.

**Comment by vshampor** [comment]
TF's mask-rcnn quantization keeps hanging. The issue is not with the quantization algorithm - hang occurs in applying the transformation layout to the model.

The transformation layout before and after the PR has the same amount of quantizers, but some of these have changed positions so as not to quantize outputs (since "quantize_outputs" is false by default).
Here's the diff of the `transformation_layout.transformations` list (left is before, right is after the PR):
```
259,263c259,263
< 0 TargetType.AFTER_LAYER rpn-box 3 0
< 0 TargetType.AFTER_LAYER rpn-box 4 0
< 0 TargetType.AFTER_LAYER rpn-box 0 0
< 0 TargetType.AFTER_LAYER rpn-box 1 0
< 0 TargetType.AFTER_LAYER rpn-box 2 0
---
> 0 TargetType.BEFORE_LAYER tf.reshape_7 0 0
> 0 TargetType.BEFORE_LAYER tf.reshape_9 0 0
> 0 TargetType.BEFORE_LAYER tf.reshape_1 0 0
> 0 TargetType.BEFORE_LAYER tf.reshape_3 0 0
> 0 TargetType.BEFORE_LAYER tf.reshape_5 0 0
310c310
< 0 TargetType.AFTER_LAYER box-predict 0 0
---
> 0 TargetType.BEFORE_LAYER tf.reshape_35 0 0
```

**Comment by vshampor** [comment]
E2E PT build: 359
E2E TF build: 366
E2E ONNX build: 356
"conformance" manual build: 58
